Agent notebook example with human feedback; Support shell command and multiple code blocks; Improve the system message for assistant agent; Improve utility functions for config lists; reuse docker image

sonichi opened the pull request:
<!-- Thank you for your contribution! Please review https://microsoft.github.io/FLAML/docs/Contribute before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

* Support shell command like "pip install"
* Support multiple code blocks like "pip install" followed by python code
* Notebook example for agent with human feedback: math and paper talk
* Further improve the messaging about LLM support
* Improve the system message for assistant agent
* Added test cases for gpt-3.5-turbo
* Added utility functions for getting config lists for a list of models.
* Commit docker image to reuse in a session.

## Related issue number

#1038 

## Checks

<!-- - I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR (note the same in integrated in our CI checks). -->
- [x] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

qingyun-wu commented on flaml/autogen/agent/coding_agent.py:
```diff
@@ -6,7 +6,11 @@
     """(Experimental) Suggest code blocks."""
 
-    DEFAULT_SYSTEM_MESSAGE = """You suggest python code (in a python coding block) for a user to execute for a given task. If you want the user to save the code in a file before executing it, put # filename: <filename> inside the code block as the first line. Finish the task smartly. Don't suggest shell command. Don't include multiple code blocks in one response. Use 'print' function for the outp ...
+    DEFAULT_SYSTEM_MESSAGE = """You are a helpful AI assistant.
```
This change requires re-running this notebook https://github.com/microsoft/FLAML/blob/main/notebook/autogen_agent_auto_feedback_from_code_execution.ipynb

sonichi replied:
done.

qingyun-wu commented:
How about renaming the `PythonAgent` to `AssistantAgent`?

gagb commented:
I ran three test using GPT-3.5 on this PR.

1. Print hello world to a file called hello.txt
2. Create a powerpoint with the text hello world in it.
3. Task 2 but with an alternative system message that allows the coding agent to generate bash commands.

Task 1 did not terminate properly.
Task 2 resulted in invalid python syntax from coding agent. Revealing that the default system message is insufficient for testing this PR (e.g., its fix to execute bash).
Task 3 again failed even though the coding agent suggested bash code blocks containing pip install commands.

I've attached the logs for all three test. The first line contains the OG instruction to the coding agent.
[test_pr1056.zip](https://github.com/microsoft/FLAML/files/11629853/test_pr1056.zip)

sonichi commented:
> I ran three test using GPT-3.5 on this PR.
> 
> 1. Print hello world to a file called hello.txt
> 2. Create a powerpoint with the text hello world in it.
> 3. Task 2 but with an alternative system message that allows the coding agent to generate bash commands.
> 
> Observations:
> 
> * Task 1 did not terminate properly.
> * Task 2 resulted in invalid python syntax from coding agent. Revealing that the default system message is insufficient for testing this PR (e.g., its fix to execute bash).
> * Task 3 again failed even though the coding agent suggested bash code blocks containing pip install commands.
> 
> I've attached the code and logs for all three tests.
> 
> See: [test_pr1056.zip](https://github.com/microsoft/FLAML/files/11629853/test_pr1056.zip)

Updated and Added test cases for gpt-3.5 based on your code in test_assistant_agent.py. Thank you very much for the input!

sonichi commented on flaml/autogen/agent/assistant_agent.py:
```diff
@@ -3,10 +3,14 @@
+    """(Experimental) Assistant agent, able to suggest code blocks."""
 
-    DEFAULT_SYSTEM_MESSAGE = """You suggest python code (in a python coding block) for a user to execute for a given task. If you want the user to save the code in a file before executing it, put # filename: <filename> inside the code block as the first line. Finish the task smartly. Don't suggest shell command. Don't include multiple code blocks in one response. Use 'print' function for the outp ...
+    DEFAULT_SYSTEM_MESSAGE = """You are a helpful AI assistant.
```
I updated the system message and passed the gpt-3.5 tests.

gagb commented:
> Updated and Added test cases for gpt-3.5 based on your code in test_assistant_agent.py. Thank you very much for the input!

Few new observations

This branch installs correctly on OSX when I used,
```pip install -e ."[autogen]"```

However the results of `test_assistant_agent.py` still look incorrect. Attaching the logs. The main issues seems to be that the assistant returns "TERMINATE" in the very first message. This may be a prompt crafting issue, but given that it works your machine but not mine is worrisome for creating a reliable agent.
[log_test_assistant_agent.txt](https://github.com/microsoft/FLAML/files/11637818/log_test_assistant_agent.txt)

kevin666aa commented on flaml/autogen/agent/assistant_agent.py:
```diff
@@ -0,0 +1,45 @@
+        "model": DEFAULT_MODEL,
+    }
+
+    def __init__(self, name, system_message=DEFAULT_SYSTEM_MESSAGE, **config):
```
I think it is better to explicitly declare all necessary configs in parameters instead of in `**config`, or to add more documents to inform what configs are needed, for example, `config_list`

sonichi replied:
Updated. Could you check whether it works?

kevin666aa commented on flaml/autogen/agent/assistant_agent.py:
```diff
@@ -0,0 +1,47 @@
+        response = oai.ChatCompletion.extract_text(responses)[0]
+        self._send(response, sender)
+
+    def reset(self):
```
If take a look at the `Agent` class in agent.py, the self._conversations is already there. Is it better to set `reset` function as a base func in `Agent` class to clear the conversation history? Also, other agents might also need the functionality of reset

kevin666aa replied:
Yes, I plan to implement an `MathUserProxyAgent` from mathchat that will count number of invalid queries. This need to be reset for a new problem.

kevin666aa replied:
I am talking about `self._conversations` in `Agent` class. It keep storing previous messages but there are no reset function for it. So I propose to put `reset` in the basic `Agent` class, then this assistant agent inherits it.

gagb commented:
> I ran three test using GPT-3.5 on this PR.
> 
> 1. Print hello world to a file called hello.txt
> 2. Create a powerpoint with the text hello world in it.
> 3. Task 2 but with an alternative system message that allows the coding agent to generate bash commands.
> 
> Observations:
> 
> * Task 1 did not terminate properly.
> * Task 2 resulted in invalid python syntax from coding agent. Revealing that the default system message is insufficient for testing this PR (e.g., its fix to execute bash).
> * Task 3 again failed even though the coding agent suggested bash code blocks containing pip install commands.
> 
> I've attached the code and logs for all three tests.
> 
> See: [test_pr1056.zip](https://github.com/microsoft/FLAML/files/11629853/test_pr1056.zip)

Current code fails to properly do pip installs. Here's the current test function and the current output.
The code executes pip installs, which succeeds. But then those installs aren't visible when it tries to execute python.
Testing done with `python:3` container.

I had issues with using conda in a miniconda docker container as well.

```
def test_gpt35(human_input_mode="NEVER", max_consecutive_auto_reply=5):
    try:
        import openai
    except ImportError:
        return
    config_list = oai.config_list_from_models(key_file_path=KEY_LOC, model_list=["gpt-3.5-turbo"])


    system_message = """You are a helpful AI assistant.
    In the following cases, suggest python code (in a python coding block) or shell script (in a sh coding block) for the user to execute. You must indicate the script type in the code block.
    1. When you need to ask the user for some info, use the code to output the info you need, for example, browse or search the web, download/read a file.
    2. When you need to perform some task with code, use the code to perform the task and output the result. Finish the task smartly. Solve the task step by step if you need to.
    If you want the user to save the code in a file before executing it, put # filename: <filename> inside the code block as the first line. Do not ask users to copy and paste the result. Instead, use 'print' function for the output when relevant. Check the execution result returned by the user.
    If the result indicates there is an error, fix the error and output the code again. Avoiding suggesting multiple code blocks. Suggeset the full code instead of partial code or code changes.
    If you use pip or conda install. Use the --quiet option to suppress the output. Make sure you execute the original code after installing the dependencies.
    Reply "TERMINATE" in the end when everything is done.
    """
    assistant = AssistantAgent(
        "coding_agent",
        request_timeout=600,
        seed=40,
        max_tokens=1024,
        config_list=config_list,
        system_message=system_message,
    )
    user = UserProxyAgent(
        "user",
        work_dir=f"{here}/test_agent_scripts",
        human_input_mode=human_input_mode,
        is_termination_msg=lambda x: x.rstrip().endswith("TERMINATE"),
        use_docker=["python:3"],
        max_consecutive_auto_reply=max_consecutive_auto_reply,
    )
    coding_task = "Print hello world to a file called hello.txt"
    assistant.receive(coding_task, user)
    coding_task = "Save a pandas df with 3 rows and 3 columns to disk."
    assistant.receive(coding_task, user)

```

[log_test_assistant_agent.txt](https://github.com/microsoft/FLAML/files/11669306/log_test_assistant_agent.txt)

thinkall commented on .github/workflows/python-package.yml:
```diff
@@ -76,8 +76,8 @@ jobs:
           pip install -e .[vw]
-      - name: Uninstall pyspark on python 3.8 and 3.9
-        if: matrix.python-version == '3.8' || matrix.python-version == '3.9'
+      - name: Uninstall pyspark on python 3.8 and 3.9 for windows
```
uninstall pyspark for other os is still needed to test on envs without pyspark
